### PR TITLE
[MRG] Early stopping now done with loss by default

### LINKED ISSUE #69: Use the training (or validation) loss value for early stopping by default
I think we should have `scoring=None` for all Gradient Boosting estimators by default and still enable early stopping.

- if `self.validation_split is not None`, call `self.loss_(y_validation, raw_predictions_validation)` instead of a `Scorer` instance.
- if `self.validation_split is None`, find a way to incrementally update an estimator of the full training loss without adding any additional computation cost (I think it should be doable but I have not checked in detail).
- if `n_iter_no_change is None or n_iter_no_change == 0` disable early stopping and go to `max_iter`.

@NicolasHug WDYT?

### PR DESCRIPTION
**Pull request by NicolasHug**
Closes #69 

Changes:

- early stopping is done iff `n_iter_no_change is not None and n_iter_no_change > 0`
- If early stopping is done:
   - Use loss iff `scoring` is None (default)
   - Use validation data iff `validation_split` is not None

There are 4 early stopping scenarios: training data and scorer, training data and loss, validation data and scorer, validation data and loss.

For furture improvements:
If early stopping is done on training data and loss, compute loss in `_update_gradients_and_hessians` to avoid multiple computations. That would also require a bit a refactoring because for now the training loss / scores are computed on the subsampled `X_binned_small` data, not on the whole data.

### CODE REVIEW

File: pygbm/gradient_boosting.py [hunk 1]
```diff
@@ -60,9 +60,9 @@ def _validate_parameters(self):
         if self.max_iter < 1:
             raise ValueError(f'max_iter={self.max_iter} must '
                              f'not be smaller than 1.')
-        if self.n_iter_no_change < 2:
+        if self.n_iter_no_change is not None and self.n_iter_no_change < 0:
             raise ValueError(f'n_iter_no_change={self.n_iter_no_change} '
-                             f'must not be smaller than 2.')
+                             f'must be positive.')
```
[the remark this answers is not in the document]
Comment (NicolasHug) [reply pygbm/gradient_boosting.py]: I think it now makes sense to allow `1` here, since the strategy has changed from before. Before #65 , setting it to `1`  would compare `scores[-1]` with itself so it didn't make sense, but now it would compare `scores[-2]` with `scores[-1]`.

[the remark this answers is not in the document]
Comment (NicolasHug) [reply pygbm/gradient_boosting.py]: But we can forbid 0 maybe?

**Comment by NicolasHug** [comment]
+1 for monitoring scores regardless of early stopping. I'm not familiar with the existing APIs for this, so I'll open an issue so that we can work on it later.

File: pygbm/gradient_boosting.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ogrisel) [reply pygbm/gradient_boosting.py]: Alright. We can keep 0 == disabled early stopping in that case.
